src/gleam/http.gleam

```diff
@@ -52,3 +65,170 @@ pub fn method_to_string(method) {
 
 pub external fn method_from_erlang(anything) -> Result(Method, Nil) =
   "gleam_http_native" "method_from_erlang"
+
+/// Data type containing an attribute value pair that constitute an HTTP header.
+pub type Header =
+  tuple(String, String)
+
+/// A general data type that can be be parameterised as a `Request` or `Response`
+pub type Message(head, body) {
+  Message(head: head, headers: List(Header), body: body)
+}
+
+/// Data structure containing the deserialized components of an HTTP request start line.
+pub type RequestHead {
+  RequestHead(
+    method: Method,
+    host: String,
+    port: Option(Int),
+    path: String,
+    query: Option(String),
+  )
+}
+
+/// Data structure containing the deserialized components of an HTTP response status line.
+pub type ResponseHead {
+  ResponseHead(status: Int)
+}
+
+/// HTTP message type parameterised as a Request.
+pub type Request(body) =
+  Message(RequestHead, body)
+
+/// HTTP message type parameterised as a Response.
+pub type Response(body) =
+  Message(ResponseHead, body)
+
+/// Construct a `Request` type.
+///
+/// The provided `uri_string` should be an absolure uri, including the scheme and host.
+/// The body type of the returned request is `Nil`,
+/// and should be set with a call to `set_body`.
+pub fn request(method: Method, uri_string: String) -> Request(Nil) {
+  let Ok(
+    uri.Uri(host: Some(host), port: port, path: path, query: query, ..),
+  ) = uri.parse(uri_string)
+  Message(
+    head: RequestHead(method, host, port, path, query),
+    headers: [],
+    body: Nil,
+  )
+}
+
+/// Return the uri that a request was sent to.
+///
+pub fn request_uri(request: Request(a), secure: Bool) -> Uri {
+  let scheme = case secure {
+    True -> "https"
+    False -> "http"
+  }
+
+  let Message(head: RequestHead(_, host, port, path, query), ..) = request
+  Uri(Some(scheme), None, Some(host), port, path, query, None)
+}
+
+/// Construct a `Response` type.
+///
+/// The body type of the returned response is `Nil`,
+/// and should be set with a call to `set_body`.
+pub fn response(status: Int) -> Response(Nil) {
+  Message(head: ResponseHead(status), headers: [], body: Nil)
+}
+
+/// Get the method of a request.
+pub fn method(message: Message(RequestHead, body)) -> Method {
+  let Message(RequestHead(method: method, ..), ..) = message
+  method
+}
+
+/// Get the host of a request.
+pub fn host(message: Message(RequestHead, body)) -> String {
+  let Message(RequestHead(host: host, ..), ..) = message
+  host
+}
+
+/// Get the port of a request, if one was explicitly given.
+pub fn port(message: Message(RequestHead, body)) -> Option(Int) {
+  let Message(RequestHead(port: port, ..), ..) = message
+  port
+}
+
+/// Return the none-empty segments of a request path.
+pub fn path_segments(message: Message(RequestHead, body)) -> List(String) {
```

Comment:
I slightly prefer `path_segments` as a path has segments but a request doesn't?

Reply:
No strong feelings here. request.segments doesn't mean anything thing else. And technically I could argue the same for query, a property of the path/url rather than the request directly.